server/src-lib/Hasura/Server/CacheUpdate.hs

```diff
@@ -0,0 +1,178 @@
+  = CacheUpdateEventLog
+  { cuelLogLevel   :: !LogLevel
+  , cuelThreadType :: !ThreadType
+  , cuelInfo       :: !Value
+  } deriving (Show, Eq)
+
+instance ToJSON CacheUpdateEventLog where
+  toJSON (CacheUpdateEventLog _ t info) =
+    object [ "thread_type" .= show t
+           , "info" .= info
+           ]
+
+instance ToEngineLog CacheUpdateEventLog where
+  toEngineLog threadLog =
+    (cuelLogLevel threadLog, "cache_update_event", toJSON threadLog)
+
+data EventPayload
+  = EventPayload
+  { _epServerId   :: !UUID
+  , _epOccurredAt :: !UTC.UTCTime
+  } deriving (Show, Eq)
+$(deriveJSON (aesonDrop 3 snakeCase) ''EventPayload)
+
+data CacheUpdateEvent
+  = CUEListenSuccess !EventPayload
+  | CUEListenFail
+  deriving (Show, Eq)
+
+instance ToJSON CacheUpdateEvent where
+  toJSON (CUEListenSuccess payload) = toJSON payload
+  toJSON CUEListenFail              = String "event listening failed"
+
+data ThreadError
+  = TEJsonParse !T.Text
+  | TEQueryError !QErr
+$(deriveToJSON
+  defaultOptions { constructorTagModifier = snakeCase . drop 2
+                 , sumEncoding = TaggedObject "type" "info"
+                 }
+ ''ThreadError)
+
+-- | An IO action that listens to postgres for events and pushes them to a Queue
+cacheUpdateEventListener
+  :: PGPool
+  -> Logger
+  -> STM.TQueue CacheUpdateEvent
+  -> IO ()
+cacheUpdateEventListener pool logger eventsQueue =
+  -- Never exits
+  forever $ do
+    listenResE <- liftIO $ runExceptT $ listen pool pgChannel notifyHandler
+    either onError return listenResE
+  where
+    threadType = TTListener
+
+    onError e = do
+      logError logger threadType $ TEQueryError e
+      -- Push a listen failed event to queue
+      STM.atomically $ STM.writeTQueue eventsQueue CUEListenFail
+      logInfo logger threadType $
+        object ["message" .= ("retrying in 10 seconds" :: T.Text)]
+      C.threadDelay $ 10 * 1000 * 1000
+
+    -- Postgres notification handler
+    notifyHandler notif = do
+      let eventChannel = PGChannel $ bsToTxt $ PQ.notifyRelname notif
+      when (eventChannel == pgChannel) $
+        case eitherDecodeStrict $ PQ.notifyExtra notif of
+          Left e -> logError logger threadType $ TEJsonParse $ T.pack e
+          Right payload -> do
+            let newEvent = CUEListenSuccess payload
+            logInfo logger threadType $ object [ "received_event" .= newEvent]
+            -- Push a success event to Queue along with event payload
+            STM.atomically $ STM.writeTQueue eventsQueue newEvent
+
+-- | An IO action that processes events from Queue
+cacheUpdateEventProcessor
+  :: PGPool
+  -> Logger
+  -> HTTP.Manager
+  -> STM.TQueue CacheUpdateEvent
+  -> IORef.IORef SchemaCache
+  -> MVar ()
+  -> UUID
+  -> UTC.UTCTime
+  -> IO ()
+cacheUpdateEventProcessor pool logger httpManager eventsQueue
+                       cacheRef lk serverId cacheInit = do
+  -- Initiate previous event IO reference
+  prevEventRef <- IORef.newIORef Nothing
```

Comment:
You don't have to store the previous event ref after the above suggested change (emit event on reconnection).